dm/syncer/data_validator_test.go

```diff
@@ -0,0 +1,460 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syncer
+
+import (
+	"context"
+	"database/sql"
+	"testing"
+	"time"
+
+	"github.com/DATA-DOG/go-sqlmock"
+	"github.com/go-mysql-org/go-mysql/mysql"
+	"github.com/go-mysql-org/go-mysql/replication"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb-tools/pkg/filter"
+	"github.com/shopspring/decimal"
+	"github.com/stretchr/testify/require"
+
+	"github.com/pingcap/tiflow/dm/dm/config"
+	"github.com/pingcap/tiflow/dm/dm/pb"
+	"github.com/pingcap/tiflow/dm/pkg/binlog"
+	"github.com/pingcap/tiflow/dm/pkg/binlog/event"
+	"github.com/pingcap/tiflow/dm/pkg/conn"
+	"github.com/pingcap/tiflow/dm/pkg/gtid"
+	"github.com/pingcap/tiflow/dm/pkg/retry"
+	"github.com/pingcap/tiflow/dm/pkg/router"
+	"github.com/pingcap/tiflow/dm/pkg/schema"
+	"github.com/pingcap/tiflow/dm/pkg/utils"
+	"github.com/pingcap/tiflow/dm/syncer/dbconn"
+)
+
+func genEventGenerator(t *testing.T) *event.Generator {
+	t.Helper()
+	previousGTIDSetStr := "3ccc475b-2343-11e7-be21-6c0b84d59f30:1-14"
+	previousGTIDSet, err := gtid.ParserGTID(mysql.MySQLFlavor, previousGTIDSetStr)
+	require.NoError(t, err)
+	latestGTIDStr := "3ccc475b-2343-11e7-be21-6c0b84d59f30:14"
+	latestGTID, err := gtid.ParserGTID(mysql.MySQLFlavor, latestGTIDStr)
+	require.NoError(t, err)
+	eventsGenerator, err := event.NewGenerator(mysql.MySQLFlavor, 1, 0, latestGTID, previousGTIDSet, 0)
+	require.NoError(t, err)
+	require.NoError(t, err)
+
+	return eventsGenerator
+}
+
+func genSubtaskConfig(t *testing.T) *config.SubTaskConfig {
+	t.Helper()
+	loaderCfg := config.LoaderConfig{
+		Dir: t.TempDir(),
+	}
+	cfg := &config.SubTaskConfig{
+		From:             config.GetDBConfigForTest(),
+		To:               config.GetDBConfigForTest(),
+		Timezone:         "UTC",
+		ServerID:         101,
+		Name:             "validator_ut",
+		ShadowTableRules: []string{config.DefaultShadowTableRules},
+		TrashTableRules:  []string{config.DefaultTrashTableRules},
+		Mode:             config.ModeIncrement,
+		Flavor:           mysql.MySQLFlavor,
+		LoaderConfig:     loaderCfg,
+		SyncerConfig: config.SyncerConfig{
+			EnableGTID: false,
+		},
+		ValidatorCfg: config.ValidatorConfig{
+			Mode:        config.ModeFull,
+			WorkerCount: 1,
+		},
+	}
+	cfg.Experimental.AsyncCheckpointFlush = true
+	cfg.From.Adjust()
+	cfg.To.Adjust()
+
+	cfg.UseRelay = false
+
+	return cfg
+}
+
+//nolint
+func genDBConn(t *testing.T, db *sql.DB, cfg *config.SubTaskConfig) *dbconn.DBConn {
+	t.Helper()
+	baseDB := conn.NewBaseDB(db, func() {})
+	baseConn, err := baseDB.GetBaseConn(context.Background())
+	require.NoError(t, err)
+	return &dbconn.DBConn{
+		BaseConn: baseConn,
+		Cfg:      cfg,
+	}
+}
+
+func Test_validator_StartStop(t *testing.T) {
```

Comment:
this is different with current test function naming convention

Reply:
it's not that easy to filter cases by struct name when using camel case, but you're right